orderer/common/cluster/block_fetcher.go

```diff
@@ -0,0 +1,486 @@
+/*
+Copyright IBM Corp. 2022 All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package cluster
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/hyperledger/fabric-protos-go/common"
+	"github.com/hyperledger/fabric-protos-go/orderer"
+	ab "github.com/hyperledger/fabric-protos-go/orderer"
+	"github.com/hyperledger/fabric/bccsp"
+	"github.com/hyperledger/fabric/common/flogging"
+	"github.com/hyperledger/fabric/common/util"
+	"github.com/hyperledger/fabric/internal/pkg/identity"
+	"github.com/hyperledger/fabric/protoutil"
+	"github.com/pkg/errors"
+	"google.golang.org/grpc"
+)
+
+//go:generate mockery -dir . -name BlockSource -case underscore -output mocks/
+
+type BlockSource interface {
+	PullBlock(seq uint64) *common.Block
+	UpdateEndpoints(endpoints []EndpointCriteria)
+	Close()
+}
+
+type attestationErrorAndResponse struct {
+	err  error
+	resp *orderer.BlockAttestationResponse
+}
+
+type attestationConnectionInfo struct {
+	stream *impatientAttestationStream
+	conn   *grpc.ClientConn
+}
+
+//go:generate mockery -dir . -name AttestationSource -case underscore -output mocks/
+type AttestationSource interface {
+	PullAttestations(seq uint64) []*orderer.BlockAttestation
+	Close()
+}
+
+type suspectSet map[string]struct{}
+
+func (s suspectSet) insert(entry string) {
+	s[entry] = struct{}{}
+}
+
+func (s suspectSet) remove(entry string) {
+	delete(s, entry)
+}
+
+func (s suspectSet) has(entry string) bool {
+	_, ok := s[entry]
+	return ok
+}
+
+type AttestationPuller struct {
+	Config FetcherConfig
+	// Internal state
+	openConnections []attestationConnectionInfo
+	cancelStream    func()
+}
+
+func (p *AttestationPuller) seekNextEnvelope(startSeq uint64) (*common.Envelope, error) {
+	return protoutil.CreateSignedEnvelopeWithTLSBinding(
+		common.HeaderType_DELIVER_SEEK_INFO,
+		p.Config.Channel,
+		p.Config.Signer,
+		nextSeekInfo(startSeq),
+		int32(0),
+		uint64(0),
+		util.ComputeSHA256(p.Config.TLSCert),
+	)
+}
+
+func (p *AttestationPuller) closeEndpoints() {
+
+	// Close the streams and grpc connections
+	for _, oc := range p.openConnections {
+		oc.stream.cancelFunc()
+		// TODO: Should we check the connection?
+		if oc.conn != nil {
+			oc.conn.Close()
+		}
+	}
+	p.openConnections = nil
+}
+
+func (p *AttestationPuller) Close() {
+	p.closeEndpoints()
+}
+
+func (p *AttestationPuller) pullAttestationBlock(ec EndpointCriteria, seq uint64) (*orderer.BlockAttestation, error) {
+	env, err := p.seekNextEnvelope(seq)
+	if err != nil {
+		return nil, err
+	}
+
+	conn, err := p.Config.Dialer.Dial(ec)
+	if err != nil {
+		return nil, err
+	}
+	defer conn.Close()
+
+	streamFunc := newImpatientStream(conn, p.Config.FetchTimeout)
```

Comment:
you have a function that returns a function and then you invoke the returned function in the next line.
Why not just have the function accept `env` as a parameter and not return a function?

Reply:
taken.